src/diagnostic/index.ts

```diff
@@ -0,0 +1,134 @@
+import path from 'path';
+
+import semver from 'semver';
+import vscode from 'vscode';
+
+import { configuration, updateConfiguration } from '../configuration';
+import { commands } from '../utils/constants';
+import { findPkgInstallDir } from '../utils/pkg';
+import { getPackageInfo } from '../utils/pkg-info';
+import { detectPm } from '../utils/pm';
+
+function isJSONParsable(str: string) {
+    try {
+        JSON.parse(str);
+    } catch {
+        return false;
+    }
+    return true;
+}
+
+export const diagnosticCollection = vscode.languages.createDiagnosticCollection('lints');
+
+export async function updateDiagnostic(document: vscode.TextDocument) {
+    if (configuration.depsVersionCheck.enable === undefined) {
+        await updateConfiguration();
+    }
+
+    if (!configuration.depsVersionCheck.enable) return;
+
+    if (
+        document.uri.fsPath.includes('node_modules') ||
+        !path.basename(document.uri.fsPath).includes('package.json') ||
+        document.languageId !== 'json'
+    )
+        return;
+
+    const text = document.getText();
+    if (!isJSONParsable(text)) return;
+
+    const diagnostics: vscode.Diagnostic[] = [];
+
+    const packageJSON = JSON.parse(text);
```

Comment:
You had done very good.
Here maybe can improve performance by avoid parse 3 times. The best way is only parse by  "jsonpos" one time. ref: https://github.com/tjx666/package-manager-enhancer/blob/main/src/utils/pkg.ts#L52